Homeworks/Homework_3/Task_3/MyThreadPool.cs

```diff
@@ -0,0 +1,210 @@
+
+                foreach (var taskContinuation in localQueue)
+                {
+                    threadPool.taskQueue.Enqueue(taskContinuation);
+                    threadPool.isTaskAvailable.Set();
+                }
+            }
+
+            public bool IsCompleted { get; private set; }
+
+            public TResult Result
+            {
+                get
+                {
+                    resultIsCalculated.WaitOne();
+                    if (caughtExcpetion != null)
+                    {
+                        throw new AggregateException(caughtExcpetion);
+                    }
+                    return result;
+                }
+            }
+
+            public IMyTask<TNewResult> ContinueWith<TNewResult>(Func<TResult, TNewResult> function)
+            {
+                if (function == null)
+                {
+                    throw new ArgumentNullException("The function is null");
+                }
+
+                lock (lockObject)
+                {
+                    if (IsCompleted)
+                    {
+                        return threadPool.Submit(() => function(Result));
+                    }
+
+                    lock (threadPool.lockObject)
+                    {
+                        if (cancellationToken.IsCancellationRequested)
+                        {
+                            throw new InvalidOperationException("Thread pool is shutdown");
+                        }
+
+                        var taskContinuation = new MyTask<TNewResult>(() => function(Result), threadPool);
+                        localQueue.Enqueue(taskContinuation.Run);
+                        Interlocked.Increment(ref threadPool.numberOfTasks);
+
+                        return taskContinuation;
+                    }
+                }
+            }
+        }
+
+        private void ExecuteTasks(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested || numberOfTasks > 0)
+            {
+                if (taskQueue.TryDequeue(out var taskRun))
+                {
+                    Interlocked.Decrement(ref numberOfTasks);
+                    taskRun();
+                }
+                else
+                {
+                    isTaskAvailable.WaitOne();
+
+                    if (cancellationToken.IsCancellationRequested || taskQueue.Count >= 2)
+                    {
+                        isTaskAvailable.Set();
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the task added to the thread pool.
+        /// </summary>
+        /// <typeparam name="TResult">Task result type.</typeparam>
+        /// <param name="function">Task function.</param>
+        public IMyTask<TResult> Submit<TResult>(Func<TResult> function)
+        {
+            if (function == null)
+            {
+                throw new ArgumentNullException("The function is null");
```

Comment:
```suggestion
                throw new ArgumentNullException(nameof(function));
```
Как раз для этого :)